designs/forLoopIteratorsDesign.md

```diff
@@ -0,0 +1,270 @@
+ 
+Input:
+```
+i : int = 0
+for i in range(10):
+    If i\>5:
+        break
+    print(i)
+else:
+    print(“i was greater than 5”) // will only work if basic strings are implemented, we can change this to print(i*100) or something
+```
+Output:
+```
+0
+1
+2
+3
+4
+5
+i was greater than 5
+```
+* **Test Case 9**: operations on range
+ 
+Input:
+```
+r : range = None
+r = range(0, 20, 2)
+print(11 in r)
+```
+Output:
+```
+False
+```
+* **Test Case 10**: inbuilt functions on range, index() returning Value Error
+ 
+Input:
+```
+r : range = None
+r = range(0, 20, 2)
+print(r.index(11))
+```
+Output:
+```
+ValueError: 11 is not in range
+```
+* **Test Case 11**:  inbuilt functions on range, index() returning index of reachable number
+ 
+Input:
+```
+r : range = None
+r = range(0, 20, 2)
+print(r.index(10))
+```
+Output:
+```
+5
+```
+* **Test Case 12**: type checking for range parameters
+ 
+Input:
+```
+range(0, 10.2, 0.2)
+```
+Output:
+```
+TypeError: float object cannot be interpreted as integer
+```
+* **Test Case 13**: type checking for range parameters
+ 
+Input:
+```
+range(0, 10, 1, 2)
+```
+Output:
+```
+TypeError: range expected at most 3 arguments, got 4
+```
+* **Test Case 14**: type checking for range parameters
+ 
+Input:
+```
+range()
+```
+Output:
+```
+TypeError: range expected 1 argument, got 0
+```
+* **Test Case 15**: type checking for range parameters
+ 
+Input:
+```
+range()
+```
+Output:
+```
+TypeError: range expected 1 argument, got 0
+```
+## Changes required in AST, IR and builtin libraries ##
+ 
+* The following changes will be required in ast.ts:
+ 
+    * In type Stmt\<A\>:
+        * { a?: Type, tag: "for", cond: Expr\<A\>, body: Array\<Stmt\<A\>\>, elseBody?: Stmt<A>}
+        * { a?: Type, tag: "break" }
+        * { a?: Type, tag: "continue" }
+    * In enum BinOp:
+        * Add a BinOP called: In
+            * Note: Thinking ahead, for the left operand of the in operator we are assuming here that an Expr can be a tuple (for loop_variable). We shall have this functionality after the group on sets/tuples/dictionaries will include it in Expr\<A\>. For now, since we are only implementing the range iterator, we shall concern ourselves with using **in** with the right operand as range only and hence, the left operand as being an id/literal which is an Expr\<A\>.
+ 
+* We shall not need any changes in ir.ts, since for, break and continue are control flow statements, and ifjmp and jmp in ir.ts cover the behavior that we need to implement these.
+ 
+* We do not need any changes in the builtin libraries. However, at a later stage of the project we might need some changes here depending on how we choose to implement general iterators
+ 
+## Adding new functions, datatypes, or files to the codebase ##
+ 
+We will not need to add any additional files or datatypes to the code base.
+ 
+For functions: we will need to add ```function rangeBuiltInFuns():``` to compilers.ts, and have functions like index() in here.
```

Comment:
Can you say more about what you expect to happen here? The function header you wrote isn't in Python _or_ TypeScript syntax, so I don't understand the goal.

If you want to add new built-in functions/classes, I would suggest thinking of how to do that with `.wat` files like in `memory.wat`.

Reply:
New design doc: https://github.com/tapan0007/chocopy-wasm-compiler-B/blob/2022/designs/forLoopIteratorsDesign.md

The builtin class's methods will be added similar to memory.wat